File: src/main/java/de/schramm/royalbash/controller/service/core/creature/GoblinRaider.java
package de.schramm.royalbash.controller.service.core.creature;

import de.schramm.royalbash.controller.service.core.Context;
import de.schramm.royalbash.controller.service.core.Creature;
import de.schramm.royalbash.controller.service.core.Game;
import lombok.Builder;
import lombok.Value;
import lombok.val;

@Value
@Builder(toBuilder = true)
public class GoblinRaider implements Creature {

    private final String id;
    private final int hitpoints;
    private final int attack;
    private final boolean isPlaceableOnSpot = true;

    @Override
    public Game invoke(Context context) {

        val game = CreatureUtil.spawnCreature(this, context);

        return game.findPlayer(context.getTargetPlayer())
                .map(player -> player.discardCards(1))
                .map(player -> game.updatePlayer(context.getTargetPlayer(), player))
                .orElse(game);
    }

    @Override
    public Creature damage(Creature attacker) {
        return this.toBuilder()
                .hitpoints(hitpoints - attacker.getAttack())
                .build();
    }

    @Override
    public boolean isDead() {
        return CreatureUtil.isDead(this);
    }
}

File: backend/src/main/kotlin/de/schramm/royalbash/domain/card/creature/GoblinRaider.kt
package de.schramm.royalbash.domain.card.creature

import de.schramm.royalbash.domain.Context
import de.schramm.royalbash.domain.Creature
import de.schramm.royalbash.domain.Game
import de.schramm.royalbash.domain.effect.DiscardHandcardsEffect
import de.schramm.royalbash.domain.spawnCreature

data class GoblinRaider(
        override val id: String,
        override val hitpoints: Int,
        override val attack: Int,
        override val cost: Int
) : Creature {

    override val name = "Goblin Raider"
    override val text = "When Goblin Raider enters the battlefield, target player discards a card."
    override val image: String? = null

    val effect = DiscardHandcardsEffect(1)

    override fun invoke(context: Context): Game {

        val updatedGame = this.spawnCreature(context)

        val updatedContext = context.copy(game = updatedGame)

        return effect.invoke(updatedContext)
    }

    override fun damage(amountOfDamage: Int): Creature =
            this.copy(hitpoints = hitpoints - amountOfDamage)
}
